src/Chainweb/Miner/RestAPI/Server.hs

```diff
@@ -131,21 +138,54 @@ solvedHandler mr (HeaderBytes hbytes) = do
     lf :: LogFunction
     lf = logFunction $ _coordLogger mr
 
-updatesHandler :: CutDb cas -> ChainBytes -> Tagged Handler Application
-updatesHandler cdb (ChainBytes cbytes) = Tagged $ \req respond -> do
+updatesHandler
+    :: Logger l
+    => MiningCoordination l cas
+    -> ChainBytes
+    -> Tagged Handler Application
+updatesHandler mr (ChainBytes cbytes) = Tagged $ \req respond -> withLimit respond $ do
     cid <- runGet decodeChainId cbytes
-    cv  <- _cut cdb >>= newIORef
-    eventSourceAppIO (go cid cv) req respond
+    cv  <- _cut (_coordCutDb mr) >>= newIORef
+    jitter <- randomRIO @Double (0.9, 1.1)
+    timer <- registerDelay (round $ jitter * realToFrac timeout * 1000000)
+    eventSourceAppIO (go timer cid cv) req respond
   where
+    timeout = _coordinationUpdateStreamTimeout $ _coordConf mr
+
     -- | A nearly empty `ServerEvent` that signals the discovery of a new
     -- `Cut`. Currently there is no need to actually send any information over
     -- to the caller.
     --
     f :: ServerEvent
     f = ServerEvent (Just $ fromByteString "New Cut") Nothing []
 
-    go :: ChainId -> IORef Cut -> IO ServerEvent
-    go cid cv = readIORef cv >>= awaitNewCutByChainId cdb cid >>= writeIORef cv >> pure f
+    go :: TVar Bool -> ChainId -> IORef Cut -> IO ServerEvent
+    go timer cid cv = do
+        c <- readIORef cv
+
+        -- await either a timeout or a new event
+        maybeCut <- atomically $ do
+            t <- readTVar timer
+            if t
+                then return Nothing
+                else Just <$> awaitNewCutByChainIdStm (_coordCutDb mr) cid c
+
+        case maybeCut of
+            Nothing -> return CloseEvent
+            Just c' -> do
+                writeIORef cv $! c'
+                return f
+
+    count = _coordUpdateStreamCount mr
+
+    withLimit respond inner = bracket
+        (atomicModifyIORef' count $ \x -> (x - 1, x - 1))
+        (const $ atomicModifyIORef' count $ \x -> (x + 1, ()))
+        (\x -> if x <= 0 then ret503 respond else inner)
+
+    ret503 respond = do
+        atomicModifyIORef' (_coord503s mr) (\c -> (c + 1, ()))
```

Comment:
```suggestion
```

incrementing the count is already taken care of in the finalizer of the `withLimit` bracket.